refactor!: new note emission API

benesjan opened the pull request:
Refactoring note emission API to use the recently introduced `MessageDelivery` global var.

```diff
- note_emission.emit(encode_and_encrypt_note(&mut context, from));
+ note_emission.emit(&mut context, from, MessageDelivery.CONSTRAINED_ONCHAIN);
```

```diff
- note_emission.emit(encode_and_encrypt_note_unconstrained(&mut context, from));
+ note_emission.emit(&mut context, from, MessageDelivery.UNCONSTRAINED_ONCHAIN);
```

```diff
- note_emission.emit(encode_and_encrypt_note_and_emit_as_offchain_message(&mut context, context.msg_sender());
+ note_emission.emit(&mut context, context.msg_sender(), MessageDelivery.UNCONSTRAINED_OFFCHAIN);
```

## Followup work

### Achieving final imports
Would give the highest priority to tackling [this issue](https://github.com/AztecProtocol/aztec-packages/issues/16771) because the import of `MessageDelivery` affects contract code and we don't want to force devs to do 2 annoying migrations. Plan on tackling that right after this PR is merged.

## Random thoughts that came to my mind while implementing this

### Consider renaming `OuterNoteEmission`
I would consider renaming `OuterNoteEmission` to `MaybeNoteEmission` because `OuterNoteEmission` is not really descriptive (it being outer feels like an implementation detail and not something you want to communicate to the struct consumer). But devs don't really get exposed to this that much so possibly doesn't matter much.

### `noteEmission.discard(...)`
I've also noticed that the discard method on the note emission structs is not ever used in our codebase. I feel like notes might be too important to ever be discarded so would revisit whether having the func makes sense.

### Emitting partial notes as offchain message
None of our partial notes implementations support emission via an offchain message.

## On using AI
In this PR there was a lot of repetitive changes in the contracts and I realized that if I write nice migration notes first then I can use that as an input to the prompt. This made the AI output flawless.

benesjan commented:
* **#16091** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16091?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16091?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

nventuro commented on noir-projects/aztec-nr/aztec/src/note/note_emission.nr:
```diff
@@ -8,15 +16,49 @@ pub struct NoteEmission<Note> {
+        context: &mut PrivateContext,
+        recipient: AztecAddress,
+    ) {
+        emit_note_in_private_log(self, context, recipient, PrivateLogContent.NO_CONSTRAINTS);
```
Why not have the enum value as a param?

benesjan replied:
Completely reworked this PR so no longer relevant

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,6 +14,7 @@ use dep::protocol_types::{
     traits::{Serialize, ToField},
 };
 
+// Note: This is now used by notes as well. TODO: Move to a shared file
```
Create an issue under [#10652](https://github.com/AztecProtocol/aztec-packages/issues/10652)

nventuro commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -80,6 +84,78 @@ where
+    }
+}
+
+// ############### What follows is the old API that is to be nuked ############### //
```
Why not remove it here? Are you just waiting to get agreement on the replacement?

benesjan replied:
Yes, wanted to get a feedback first as the changes will be all over the place (it's in the PR description 😝).

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -25,17 +26,6 @@ fn assert_note_exists(context: PrivateContext, note_hash_counter: u32) {
     assert(note_exists, "Can only emit a note log for an existing note.");
 }
 
-pub fn compute_note_log<Note>(
```
This function became redundant because now we have only 1 note emission func and hence it had only 1 callsite.

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -44,20 +34,24 @@ pub fn compute_partial_note_log<Note>(
     recipient: AztecAddress,
     msg_type: u64,
-) -> [Field; PRIVATE_LOG_SIZE_IN_FIELDS]
+) -> [Field; PRIVATE_LOG_CIPHERTEXT_LEN]
```
Feels like we will need to fix the logs vs message naming because in the past we had no offchain messages and now we have incorrect naming because this ciphertext goes to message as well. Hence the `PRIVATE_LOG_CIPHERTEXT_LEN` constant name is a kinda stale.

Actually not super sure if it's stale because the private logs are what places the size constraint here and we use the same size for messages as well just because it's convenient. Probably best to not deal with now and see what the future brings.

nventuro replied:
>Actually not super sure if it's stale because the private logs are what places the size constraint here and we use the same size for messages as well just because it's convenient.

We should probably do something like `CIPHERTEXT_LEN = PRIVATE_LOG_LEN - MESSAGE_OVERHEAD` or w/e.

benesjan replied:
This is what it's currently defined:

`pub global PRIVATE_LOG_CIPHERTEXT_LEN: u32 = PRIVATE_LOG_SIZE_IN_FIELDS - 1; // -1 for the tag`

The "-1" is for the tag which does not correspond to `MESSAGE_OVERHEAD` (it's tagging related so only relevant to private logs).

But agree that it makes sense to rename `PRIVATE_LOG_CIPHERTEXT_LEN`. WDYT about calling that instead `MESSAGE_CIPHERTEXT_LEN` or just `MESSAGE_LEN`? My mental model is that message is the thing that can be emitted either in private log or as an offchain effect/offchain message. Do you think this mental model is correct?

UPDATE:
Just notice we have this in the codebase:

```
        #[$utility]
        unconstrained fn process_message(
            **message_ciphertext**: BoundedVec<Field, aztec::protocol_types::constants::**PRIVATE_LOG_CIPHERTEXT_LEN**>,
            message_context: aztec::messages::processing::message_context::MessageContext,
        ) {
            aztec::messages::discovery::process_message::do_process_message(
                context.this_address(),
                _compute_note_hash_and_nullifier,
                message_ciphertext,
                message_context,
            );
        }
```

`MESSAGE_CIPHERTEXT_LEN` makes sense.

@nventuro agree with the renaming?

benesjan replied:
Since these renamings are quick to implement and I've already had [this issue](https://github.com/AztecProtocol/aztec-packages/issues/15012) created a while ago decided to just had a stab at it in [this PR](https://github.com/AztecProtocol/aztec-packages/pull/16882).

benesjan commented on noir-projects/aztec-nr/value-note/src/utils.nr:
```diff
@@ -33,7 +33,7 @@ pub fn increment(
     let note = ValueNote::new(amount, recipient);
     // Insert the new note to the owner's set of notes and emit the log if value is non-zero.
     // TODO: consider how to give devs choice over this `encode_and_encrypt_note` function. At the moment, it's hard-coded as one of many potential strategies.
```
@iAmMichaelConnor you originally placed this TODO here in [this PR](https://github.com/AztecProtocol/aztec-packages/pull/11503).

It became very relevant now as now we can easily achieve that by having `delivery_mode: u8` as arg of the `increment` function.

Seems like a very cute and exciting outcome of this PR. Nico very smort for coming up with this approach. ~Would do that in a followup PR if everyone agrees.~

Unbothered. Moisturized. Happy. In My Lane. Focused. Flourishing.

## Update
The `increment(...)` and `decrement(...)` methods are only used in `Benchmarking` and `StatefulTest` contracts. I would replace its use and nuke the whole `noir-projects/aztec-nr/value-note/src/utils.nr` file. It feels like an ugly abstraction that shouldn't be in a package called `value-note`. Agreed?

nventuro replied:
Alternatively this could return the `NoteEmission`. But you see how at that point this function is not really doing anything.

I think it might be ok to delete these yea.

benesjan replied:
Created [an issue](https://github.com/AztecProtocol/aztec-packages/issues/16879) for this.

benesjan commented on boxes/boxes/react/src/contracts/src/main.nr:
```diff
@@ -1,14 +1,14 @@
-use dep::aztec::macros::aztec;
+use aztec::macros::aztec;
```
I was doing these refactors with AI and since it was already touching imports told it to improve style of those. Since it's just a style of imports thought it's fine even though the change is unrelated.

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/logs/note.nr:
```diff
@@ -73,98 +67,58 @@ where
+/// * `recipient` - The address that should receive this note
+/// * `delivery_mode` - Controls encryption, tagging, and delivery constraints. Must be a compile-time constant.
+///   See `MessageDeliveryEnum` for details on the available modes.
+pub fn emit_note<Note>(
```
This function now has only 1 callsite and that is in the emit function of the NoteEmission struct.

I would just move the code in that function in a followup PR. Would the reviewer agree?

Didn't do it it in this PR as I wanted feedback on it and it makes it easier to review to keep the stuff in the same places here.

nventuro replied:
Possibly. I'd take a look at where everything lives and figure out if we also want to move other stuff etc. As mentioned by you above there's already some other weird stuff.

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16889

nventuro commented on docs/docs/migration_notes.md:
```diff
@@ -45,6 +45,46 @@ As we prepare for a bigger `Wallet` interface refactor and the upcoming `WalletS
+1. `encode_and_encrypt_note_and_emit_as_offchain_message` has been removed in favor of using `emit` with `MessageDelivery.UNCONSTRAINED_OFFCHAIN`
+1. Note emission now takes a `delivery_mode` parameter with the following values:
+   - `CONSTRAINED_ONCHAIN`: For on-chain delivery with cryptographic guarantees that recipients can discover and decrypt messages. Uses constrained encryption but is slower to prove. Best for critical messages that contracts need to verify.
+   - `UNCONSTRAINED_ONCHAIN`: For on-chain delivery without encryption constraints. Faster proving but trusts the sender. Good when recipients can verify message validity through other means.
```
>Good when recipients can verify message validity through other means.

I think you want to say something more like 'good when the sender is forced to perform encryption correctly', e.g. because they're buying something and will only get it if the recipient sees the note.

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16091/commits/ae7a53eac47f5fd70587dcb05c063bec27ef1019

nventuro commented on docs/docs/migration_notes.md:
```diff
@@ -45,6 +45,46 @@ As we prepare for a bigger `Wallet` interface refactor and the upcoming `WalletS
+1. Note emission now takes a `delivery_mode` parameter with the following values:
+   - `CONSTRAINED_ONCHAIN`: For on-chain delivery with cryptographic guarantees that recipients can discover and decrypt messages. Uses constrained encryption but is slower to prove. Best for critical messages that contracts need to verify.
+   - `UNCONSTRAINED_ONCHAIN`: For on-chain delivery without encryption constraints. Faster proving but trusts the sender. Good when recipients can verify message validity through other means.
+   - `UNCONSTRAINED_OFFCHAIN`: For off-chain delivery (e.g. cloud storage) without constraints. Lowest cost since no on-chain storage needed. Requires custom infrastructure for delivery.
```
I'd consider adding in the last two that there are no guarantees at all on the delivery, and that the recipient may never find the notes.

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16091/commits/a6d2ea7a4d60cd226857761ee94a659151921dcc

nventuro reviewed (commented):
>I would consider renaming OuterNoteEmission to MaybeNoteEmission because OuterNoteEmission is not really descriptive (it being outer feels like an implementation detail and not something you want to communicate to the struct consumer). But devs don't really get exposed to this that much so possibly doesn't matter much.

OptionNoteEmission would be ok, yes. It only exists because none of the `Option` methods are a good fit - we'd want something like `and_then` but with no chaining.

>noteEmission.discard(...)
>
>I've also noticed that the discard method on the note emission structs is not ever used in our codebase. I feel like notes might be too important to ever be discarded so would revisit whether having the func makes sense.

We could add a `discard` example I guess. It'd come up in transiento reads, e.g. 

```noir
let note = priv_mut.get_note().discard(); // discard as it's about to get deleted so no point in emitting anything
priv_mut.replace(Note::new(note + 1)).emit(...);
```

>Emitting partial notes as offchain message
>None of our partial notes implementations support emission via an offchain message.

Would be good to address this, yes.

nventuro reviewed (approved):
Lovely work!

benesjan committed:
refactor!: new note emission API

Refactoring note emission API to use the recently introduced `MessageDelivery` global var.

```diff
- note_emission.emit(encode_and_encrypt_note(&mut context, from));
+ note_emission.emit(&mut context, from, MessageDelivery.CONSTRAINED_ONCHAIN);
```

```diff
- note_emission.emit(encode_and_encrypt_note_unconstrained(&mut context, from));
+ note_emission.emit(&mut context, from, MessageDelivery.UNCONSTRAINED_ONCHAIN);
```

```diff
- note_emission.emit(encode_and_encrypt_note_and_emit_as_offchain_message(&mut context, context.msg_sender());
+ note_emission.emit(&mut context, context.msg_sender(), MessageDelivery.UNCONSTRAINED_OFFCHAIN);
```

## Followup work

### Achieving final imports
Would give the highest priority to tackling [this issue](https://github.com/AztecProtocol/aztec-packages/issues/16771) because the import of `MessageDelivery` affects contract code and we don't want to force devs to do 2 annoying migrations. Plan on tackling that right after this PR is merged.

## Random thoughts that came to my mind while implementing this

### Consider renaming `OuterNoteEmission`
I would consider renaming `OuterNoteEmission` to `MaybeNoteEmission` because `OuterNoteEmission` is not really descriptive (it being outer feels like an implementation detail and not something you want to communicate to the struct consumer). But devs don't really get exposed to this that much so possibly doesn't matter much.

### `noteEmission.discard(...)`
I've also noticed that the discard method on the note emission structs is not ever used in our codebase. I feel like notes might be too important to ever be discarded so would revisit whether having the func makes sense.

### Emitting partial notes as offchain message
None of our partial notes implementations support emission via an offchain message.

## On using AI
In this PR there was a lot of repetitive changes in the contracts and I realized that if I write nice migration notes first then I can use that as an input to the prompt. This made the AI output flawless.
